sw/Core/Src/touch.h

```diff
@@ -545,148 +547,338 @@ s16 midi_chan_pitchbend[16];
+		synthfingerdown_nogatelen_internal |= bit;
+	}
+	else {
+		synthfingerdown_nogatelen_internal &= ~bit;
 	}
 
-	/////////////////// 
-
-	synth_dst_finger->pressure = p;
-//	int total_ui_pressure_before_me = total_ui_pressure;
-	total_ui_pressure += maxi(0, p);
-
-	//if (uif->written) // written is only set if we are confident to update the position
-		synth_dst_finger->pos = uif->pos;
-	//else
-	//	synth_dst_finger->pos = synth_prev_finger->pos;
+	// === SEQ PLAYING === //
+
+	// sequencer overrides any touch or midi input, when it is playing and data is present for the finger on this step
+
+	if (isplaying() || seq_rhythm.did_a_retrig) {
+		FingerRecord *seq_record = readpattern(fi);
+		if (seq_record) {
+            int seq_pressure = seq_record->pressure[substep] * 12;
+            int seq_position = seq_record->pos[substep / 2];
+			int gatelen = param_eval_finger(P_GATE_LENGTH, fi, synth_finger) >> 8;
+			int small_substep = calcseqsubstep(0, 256);
+
+            seq_pressure = (seq_pressure && !seq_rhythm.supress && small_substep <= gatelen) ? randrange(seq_pressure, seq_pressure+12)-512 : 0;
+            if (seq_pressure > 0) {
+	            read_from_seq = true;
+                pressure = seq_pressure;
+                position = randrange(seq_position * 8, seq_position * 8 + 8);
+
+				// override touch and midi data
+				midi_pressure_override &= ~bit;
+				midi_pitch_override &= ~bit;
+				synth_finger->pressure = pressure;
+				total_ui_pressure += maxi(0, pressure);
+				synth_finger->pos = position;
+				if (synth_finger->pressure > 0) {
+					synthfingerdown_nogatelen_internal |= bit;
+				}
+				else {
+					synthfingerdown_nogatelen_internal &= ~bit;
+				}
+	        }
+		}
+		// mute sequencer while pressing clear
+		if (shift_down == SB_CLEAR)
+			synth_finger->pressure = 0;
+	}
 
-	int phase0 = calcseqsubstep(0, 8);
-	if (fi==0)
-		read_from_seq = false;
-	bool any_fingers_down = (total_ui_pressure > 0 || prev_total_ui_pressure > 0);
-	if (fi==7 && shift_down!=SB_CLEAR && total_ui_pressure > 0 && total_ui_pressure >= prev_total_ui_pressure) {
-		// update latch from ui state
-		for (int fi=0;fi<8;++fi) {
-			u8 maxpos = 0, minpos = 255, maxpressure = 0;
-			Finger* f = fingers_synth_time[fi];
-			for (int j = 0; j < 8; ++j, ++f) {
-				u8 p = clampi((f->pos + 4) / 8, 0, 255);
-				minpos = mini(p, minpos);
-				maxpos = maxi(p, maxpos);
-				u8 pr = clampi(f->pressure / 12, 0, 255);
-				maxpressure = maxi(maxpressure, pr);
-			}
-			//if (avgvel > latch[fi].avgvel+1 || pressure_not_rapidly_decreasing)
-			{
-//				if (fi == 4) EmuDebugLog("latch pressure %d, tot %d prev %d\n", maxpressure, total_ui_pressure, prev_total_ui_pressure);
-				latch[fi].avgvel = maxpressure;
-				latch[fi].minpos = minpos;
-				latch[fi].maxpos = maxpos;
+	// RJ: This bit shifts some values around but I'm not sure what they are doing exactly?
+	// It XORs the last two bits with the new position? Why? Randomization?
```

Comment:
so from what I recall, we are holding a rolling buffer of finger positions, that have natural 'noise' to them and this directly affects the detuning / unison of the voice, and varies depends on how much movement there is in your finger. I wanted to make it so that when you first press a new note, *all* the old position values get rewritten with the new position, to make it less sluggish; however, I also didnt want to force them all to be identical values because that would lead to an unnatural moment of extreme unison. so, I preserved the bottom 2 bits of position to give a bit of organic-ness in the overridden history.